DOC-848: Document the new `language` UI in 5.9 (#2015)
tinymce/tinymce-docs: _includes/configuration/content-langs.md

## `content_langs`

{{ site.requires_5_9v }}

This option sets the languages displayed in the `language` dropdown toolbar button and the `language` menu item. The `content_langs` option has no default, and if it is not set then the `language` toolbar button and menu item will not be available. The option accepts an array of language objects.

Each language object is a JavaScript object defined using the following properties.

| Name | Type | Requirement | Description |
| ---- | ---- | ----------- | ----------- |
| title | String | required | The name shown for this language in the UI. |
| code | String | required | The language tag, as defined in [Tags for Identifying Languages (BCP47)](https://www.ietf.org/rfc/bcp/bcp47.txt), for this language. This tag is placed in the `lang` attribute for elements formatted using this language. |

> | customCode | String | optional | An alternative language tag, that does not need to be compatible with BCP47. This code is placed in the `data-mce-lang` attribute for elements formatted using this language, if it is defined. |

Author: No it won't, we checked. It's only `data-mce-bogus` things that get removed like that.

**Type:** `Array`

### Example: using `content_langs`

```js
tinymce.init({
    selector: 'textarea', // change this according to your HTML
    toolbar: 'language',
    content_langs: [
        { title: 'English', code: 'en' },
        { title: 'Spanish', code: 'es' },
        { title: 'French', code: 'fr' },
        { title: 'German', code: 'de' },
        { title: 'Portuguese', code: 'pt' },
        { title: 'Chinese', code: 'zh' }
    ]
});
```

### Using `content_langs` with Spell Checker Pro

> The Spell Checker Pro plugin also defines a `language` toolbar button, with similar functionality. In {{site.productname}} 5.9 and later, the `language` toolbar button and menu item replace this. The `content_langs` option has precedence over [`spellchecker_select_languages`]({{site.baseurl}}/plugins/premium/tinymcespellchecker/#spellchecker_select_languages) for determining which languages are available, where both are defined. Where the `content_langs` option is not specified, the `spellchecker_select_languages` option or its default will set the available languages.

Author: Gave this a rewrite, let me know what you think.

> The Spell Checker Pro plugin also defines a `language` toolbar button, with similar functionality. In {{site.productname}} 5.9 and later, the `language` toolbar button and menu item replace this. The `content_langs` option has precedence over [`spellchecker_select_languages`]({{site.baseurl}}/plugins/premium/tinymcespellchecker/#spellchecker_select_languages) for determining which languages are available, where both are defined. Where the `content_langs` option is not specified, the `spellchecker_select_languages` option or its default will set the available languages.

Author: Replaced this with just a hint saying "btw we use the deprecated setting value as a fallback"

> Content languages applied by the `language` toolbar button or menu item will determine which dictionary Spell Checker Pro uses to check words within those elements. Ensure that the `code` specified is within the [Supported Languages]({{site.baseurl}}/plugins/premium/tinymcespellchecker/#supportedlanguages) section. To use a language code that is not valid according to [Tags for Identifying Languages (BCP47)](https://www.ietf.org/rfc/bcp/bcp47.txt), set `customCode` to the code supported by Spell Checker Pro and set `code` to an equivalent code that is valid as per BCP47.

Author: Gave this a rewrite, let me know what you think

| customCode | String | optional | An alternative language tag, that does not need to be compatible with BCP47. This code is placed in the `data-mce-lang` attribute for elements formatted using this language, if it is defined. |

**Type:** `Array<ContentLang>`

**Default Value:**
```js
[
  { title: 'English', code: 'en' },
  { title: 'Spanish', code: 'es' },
  { title: 'French', code: 'fr' },
  { title: 'German', code: 'de' },
  { title: 'Portuguese', code: 'pt' },
  { title: 'Chinese', code: 'zh' }
]
```

### Example: using `content_langs`

```js
tinymce.init({
    selector: 'textarea', // change this according to your HTML
    toolbar: 'language',
    content_langs: [
        { title: 'English', code: 'en' },
        { title: 'Spanish', code: 'es' },
        { title: 'French', code: 'fr' },
        { title: 'German', code: 'de' },
        { title: 'Portuguese', code: 'pt' },
        { title: 'Chinese', code: 'zh' }
    ]
});
```

### Using `content_langs` with Spellchecker Pro

> The Spellchecker Pro plugin also defines a `language` toolbar button, with similar functionality. In TinyMCE 5.9 and later, the `content_langs` setting has precedence over [`spellchecker_select_languages`]({{site.baseurl}}//plugins/premium/tinymcespellchecker/#spellchecker_select_languages) and the `language` toolbar button will be controlled by the `content_langs` setting if both are specified.

Author: ```suggestion
The Spellchecker Pro plugin also defines a `language` toolbar button, with similar functionality. In TinyMCE 5.9 and later, the `content_langs` option has precedence over [`spellchecker_select_languages`]({{site.baseurl}}//plugins/premium/tinymcespellchecker/#spellchecker_select_languages) and the `language` toolbar button will be controlled by the `content_langs` option if both are specified.
```

Content languages applied by the `language` toolbar button or menu item will determine which dictionary Spellchecker Pro uses to check words within those elements. Ensure that the `code` you specify is within the [Supported Languages]({{site.baseurl}}//plugins/premium/tinymcespellchecker/#supportedlanguages) section. If the language code you with to use is not valid according to [Tags for Identifying Languages (BCP47)](https://www.ietf.org/rfc/bcp/bcp47.txt), set `customCode` to the code supported by Spellchecker Pro and set `code` to an equivalent code that is valid as per BCP47.

```js
tinymce.init({
    selector: 'textarea', // change this according to your HTML

>     toolbar: 'language',

Author: ```suggestion
    toolbar: 'language',
    plugins: 'tinymcespellchecker',
```

> ## `content_langs`

Reviewer: ```suggestion
## content_langs
```
Not sure on this one, as some files have it in code but others don't. I believe this shouldn't be marked up as code though.

Author: Might see what Tyler says on this for now.

{{ site.requires_5_9v }}

> This option allows you to override the languages displayed in the `language` dropdown toolbar button and the `language` menu item. This option accepts an array of language objects.

Reviewer: We shouldn't be using "you" here. See https://www.tiny.cloud/docs/configure/contributing-docs/#pronouns

Each language object is a JavaScript object defined using the following properties.

| Name | Type | Requirement | Description |
| ---- | ---- | ----------- | ----------- |
| title | String | required | The name shown for this language in the UI. |
| code | String | required | The language tag, as defined in [Tags for Identifying Languages (BCP47)](https://www.ietf.org/rfc/bcp/bcp47.txt), for this language. This tag is placed in the `lang` attribute for elements formatted using this language. |
| customCode | String | optional | An alternative language tag, that does not need to be compatible with BCP47. This code is placed in the `data-mce-lang` attribute for elements formatted using this language, if it is defined. |

> **Type:** `Array<ContentLang>`

Reviewer: This should just be Array, as the docs only use simple types.
```suggestion
**Type:** `Array`
```

> **Default Value:**

Reviewer: This information is no longer correct as was mentioned on Wednesday. This will not have a default moving forward, so this needs changing to reflect that.

```js
[
  { title: 'English', code: 'en' },
  { title: 'Spanish', code: 'es' },
  { title: 'French', code: 'fr' },
  { title: 'German', code: 'de' },
  { title: 'Portuguese', code: 'pt' },
  { title: 'Chinese', code: 'zh' }
]
```

### Example: using `content_langs`

```js
tinymce.init({
    selector: 'textarea', // change this according to your HTML
    toolbar: 'language',

>     content_langs: [

Reviewer: An example should always use something different to the default.

Author: There no longer is a default (unless you're using Spell Checker Pro, in which case the default is equivalent to this but it isn't actually this) so is that okay now?

        { title: 'English', code: 'en' },
        { title: 'Spanish', code: 'es' },
        { title: 'French', code: 'fr' },
        { title: 'German', code: 'de' },
        { title: 'Portuguese', code: 'pt' },
        { title: 'Chinese', code: 'zh' }
    ]
});
```

> ### Using `content_langs` with Spellchecker Pro

Reviewer: ```suggestion
### Using `content_langs` with Spell Checker Pro
```

> The Spellchecker Pro plugin also defines a `language` toolbar button, with similar functionality. In TinyMCE 5.9 and later, the `content_langs` option has precedence over [`spellchecker_select_languages`]({{site.baseurl}}//plugins/premium/tinymcespellchecker/#spellchecker_select_languages) and the `language` toolbar button will be controlled by the `content_langs` option if both are specified.

Reviewer: ```suggestion
The Spellchecker Pro plugin also defines a `language` toolbar button, with similar functionality. In {{site.product name}} 5.9 and later, the `content_langs` option has precedence over [`spellchecker_select_languages`]({{site.baseurl}}/plugins/premium/tinymcespellchecker/#spellchecker_select_languages) and the `language` toolbar button will be controlled by the `content_langs` option if both are specified.
```
Typo and we should be using the product name variable instead.

> Content languages applied by the `language` toolbar button or menu item will determine which dictionary Spellchecker Pro uses to check words within those elements. Ensure that the `code` you specify is within the [Supported Languages]({{site.baseurl}}//plugins/premium/tinymcespellchecker/#supportedlanguages) section. If the language code you with to use is not valid according to [Tags for Identifying Languages (BCP47)](https://www.ietf.org/rfc/bcp/bcp47.txt), set `customCode` to the code supported by Spellchecker Pro and set `code` to an equivalent code that is valid as per BCP47.

Reviewer 2: "you" needs to be removed here too.

```suggestion
Content languages applied by the `language` toolbar button or menu item will determine which dictionary Spell Checker Pro uses to check words within those elements. Ensure that the `code` you specify is within the [Supported Languages]({{site.baseurl}}//plugins/premium/tinymcespellchecker/#supportedlanguages) section. If the language code you wish to use is not valid according to [Tags for Identifying Languages (BCP47)](https://www.ietf.org/rfc/bcp/bcp47.txt), set `customCode` to the code supported by Spell Checker Pro and set `code` to an equivalent code that is valid as per BCP47.
```

| `customCode` | String | optional | An alternative language tag, that does not need to be compatible with BCP47. This string is placed in the `data-mce-lang` attribute for elements formatted using this language, if it is defined. |

{% if page.name != "tinymcespellchecker.md" %}
There is no default value for the `content_langs` option. If no value is specified, the `language` toolbar button and menu item are not available. A default value for this option is provided by the Spell Checker Pro plugin. For information on using the `content_langs` option with the Spell Checker Pro plugin, see: [Spell Checker Pro plugin - `content_langs`]({{site.baseurl}}/plugins/premium/tinymcespellchecker/#content_langs).
{% endif %}

**Type:** `Array`

{% if page.name == "tinymcespellchecker.md" %}
**Default Value:**

```js
[
    { title: 'English', code: 'en' },
    { title: 'Spanish', code: 'es' },
    { title: 'French', code: 'fr' },
    { title: 'German', code: 'de' },
    { title: 'Portuguese', code: 'pt' },
    { title: 'Chinese', code: 'zh' }
]
```

> > **Note:** This default is provided by the Spell Checker Pro plugin, it is not available in the core editor. The languages specified by the deprecated [`spellchecker_select_language`](#spellchecker_select_language) option are used as a fallback if they are available.

Reviewer: This note still seems out of place. I'm wondering if it should be above the `Type` as well?

Reviewer: Actually... reading more, this is **not** a SpellChecker pro setting, so why is it even shown in the spellcheck pro page? This just needs to be linked from the `spellchecker_language_select` setting with the default mentioned there instead.

- The `code` property is valid language tag, according to [Tags for Identifying Languages (BCP47)](https://www.ietf.org/rfc/bcp/bcp47.txt)
- Either the `code` property is one of the [Supported Languages](#supportedlanguages) or the `customCode` override property is defined.
- The `customCode` property is one of the [Supported Languages](#supportedlanguages) if it is defined.

For example:

```js
tinymce.init({
    selector: 'textarea', // change this according to your HTML
    toolbar: 'language',
    plugins: 'tinymcespellchecker',
    content_langs: [
        { title: 'English (US)', code: 'en_US' },
        { title: 'English (US Medical)', code: 'en_US', customCode: 'en_US-medical' },
        { title: 'English (UK)', code: 'en_UK' },
        { title: 'English (UK Medical)', code: 'en_UK', customCode: 'en_UK-medical' },
        { title: 'Spanish', code: 'es' },
        { title: 'French', code: 'fr' },
        { title: 'German', code: 'de' },
        { title: 'Portuguese', code: 'pt' },
        { title: 'Chinese', code: 'zh' }
    ]
});
```
{% else %}
### Example: Using `content_langs`

```js
tinymce.init({

> selector: 'textarea', // change this according to your HTML
> toolbar: 'language',
> content_langs: [
>     { title: 'English', code: 'en' },
>     { title: 'Spanish', code: 'es' },
>     { title: 'French', code: 'fr' },
>     { title: 'German', code: 'de' },
>     { title: 'Portuguese', code: 'pt' },
>     { title: 'Chinese', code: 'zh' }
>     ]

Reviewer: ```suggestion
  selector: 'textarea', // change this according to your HTML
  toolbar: 'language',
  content_langs: [
    { title: 'English', code: 'en' },
    { title: 'Spanish', code: 'es' },
    { title: 'French', code: 'fr' },
    { title: 'German', code: 'de' },
    { title: 'Portuguese', code: 'pt' },
    { title: 'Chinese', code: 'zh' }
  ]
```
Formatting issues here... I'd highly recommend reading this on launcher or using the local demo, as it would help catch a lot of these visual issues.

{% if page.name == "tinymcespellchecker.md" %}
### `content_langs`
{% else %}
## `content_langs`
{% endif %}

{{ site.requires_5_9v }}

> This option sets the languages displayed in the `language` dropdown toolbar button and the `language` menu item. The option accepts an array of language objects. Each language object is a JavaScript object defined using the following properties.

Reviewer: ```suggestion
This option sets the languages displayed in the `language` toolbar button and the `language` menu item. The option accepts an array of language specs. Each language spec is a JavaScript object defined using the following properties.
```
Dropdown is necessary here. We're also repeating `objects` too much here and it could be confusing since it's worded in such a way they are different, so tried to come up with another name.

| Name | Type | Requirement | Description |
| ---- | ---- | ----------- | ----------- |
| `title` | String | required | The name shown for this language in the UI. |

> | `code` | String | required | The language tag, as defined in [Tags for Identifying Languages (BCP47)](https://www.ietf.org/rfc/bcp/bcp47.txt), for this language. This tag is placed in the `lang` attribute for elements formatted using this language. |

Reviewer: `tag` could be confusing here, as tag often refers to the element name/type. Can we please come up with another term here please?

Note: I know this is the phrase used in the RFC, but most customers aren't going to be proficient in RFC specs, so it needs to be dumbed down a little. At the very least we should do what [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/lang) does and use "language tag" in quotes and never use "tag" standalone.

| `customCode` | String | optional | An alternative language tag, that does not need to be compatible with BCP47. This code is placed in the `data-mce-lang` attribute for elements formatted using this language, if it is defined. |

**Type:** `Array`

{% if page.name == "tinymcespellchecker.md" %}

> **Default:**

Reviewer: ```suggestion
**Default Value:**
```
The heading used elsewhere is this, not `Default`.

```js
[
    { title: 'English', code: 'en' },
    { title: 'Spanish', code: 'es' },
    { title: 'French', code: 'fr' },
    { title: 'German', code: 'de' },
    { title: 'Portuguese', code: 'pt' },
    { title: 'Chinese', code: 'zh' }
]
```

> **Note:** This default is provided by the Spell Checker Pro plugin, it is not available in the core editor.

> > **Note:** The languages specified by the deprecated [`spellchecker_select_language`](#spellchecker_select_language) option are used as a fallback if they are available.

Reviewer: Having 2 notes like this reads and looks very strange. Can we merge this into the one note, as it's all related?

{% else %}

> There is no default value for the `content_langs` option. If no value is specified, the `language` toolbar button and menu item are not available. A default value for this option is provided by the Spell Checker Pro plugin. For information on using the `content_langs` option with the Spell Checker Pro plugin, see: [Spell Checker Pro plugin - `content_langs`]({{site.baseurl}}/plugins/premium/tinymcespellchecker/#content_langs).

Reviewer: This whole section reads really oddly when read in the rendered view, as it just feels as though it's coming out of nowhere... This really should be before we specify the `type` as nowhere else in the docs do we include additional information after the `type`.

I wonder if the structure should be something like this instead, as that's similar to what we do in the UI components pages:

- Intro para
- this para
- type
- Spec heading and table
- example

Author: I've moved this paragraph earlier, but I don't think putting it before the type / spec table is going to work without some heavier reworking. Let me know if you still find this flow hard to follow, and I'll shuffle things again

{% if page.name == "tinymcespellchecker.md" %}
### `content_langs`
{% else %}
## `content_langs`
{% endif %}

{{ site.requires_5_9v }}

This option sets the languages displayed in the `language` dropdown toolbar button and the `language` menu item. The option accepts an array of language objects. Each language object is a JavaScript object defined using the following properties.

| Name | Type | Requirement | Description |
| ---- | ---- | ----------- | ----------- |
| `title` | String | required | The name shown for this language in the UI. |
| `code` | String | required | The language tag, as defined in [Tags for Identifying Languages (BCP47)](https://www.ietf.org/rfc/bcp/bcp47.txt), for this language. This tag is placed in the `lang` attribute for elements formatted using this language. |
| `customCode` | String | optional | An alternative language tag, that does not need to be compatible with BCP47. This code is placed in the `data-mce-lang` attribute for elements formatted using this language, if it is defined. |

**Type:** `Array`

{% if page.name == "tinymcespellchecker.md" %}
**Default:**

> ```js

Reviewer 3: ```suggestion

```js
```

[
    { title: 'English', code: 'en' },
    { title: 'Spanish', code: 'es' },
    { title: 'French', code: 'fr' },
    { title: 'German', code: 'de' },
    { title: 'Portuguese', code: 'pt' },
    { title: 'Chinese', code: 'zh' }
]
```

> > **Note:** This default is provided by the Spell Checker Pro plugin, it is not available in the core editor.

Reviewer 3: ```suggestion

> **Note:** This default is provided by the Spell Checker Pro plugin, it is not available in the core editor.
```

> **Note:** The languages specified by the deprecated [`spellchecker_select_language`](#spellchecker_select_language) option are used as a fallback if they are available.

{% else %}

> There is no default value for the `content_langs` option. If no value is specified, the `language` toolbar button and menu item are not available. A default value for this option is provided by the Spell Checker Pro plugin. For more information, see: [Spell Checker Pro plugin - `content_langs`]({{site.baseurl}}/plugins/premium/tinymcespellchecker/#content_langs).

Reviewer 3: ```suggestion
There is no default value for the `content_langs` option. If no value is specified, the `language` toolbar button and menu item are not available. A default value for this option is provided by the Spell Checker Pro plugin. For information on using the `content_langs` option with the Spell Checker Pro plugin, see: [Spell Checker Pro plugin - `content_langs`]({{site.baseurl}}/plugins/premium/tinymcespellchecker/#content_langs).
```
